src/lib/lspace/elements/element.rs

```diff
@@ -7,112 +7,93 @@ use layout::lreq::LReq;
+
+
+pub struct ElementRef {
     x: Rc<RefCell<Box<TElement>>>
 }
 
-
-impl ElementChildRef {
-    pub fn new<T: TElement + 'static>(x: T) -> ElementChildRef {
-        return ElementChildRef{x: Rc::new(RefCell::new(Box::new(x)))};
+impl ElementRef {
+    pub fn new<T: TElement + 'static>(x: T) -> ElementRef {
+        return ElementRef{x: Rc::new(RefCell::new(Box::new(x)))};
     }
 
-    pub fn get(&self) -> Ref<Box<TElement>> {
+    pub fn get(&self) -> ElemBorrow {
         return self.x.borrow();
     }
 
-    pub fn get_mut(&mut self) -> RefMut<Box<TElement>> {
+    pub fn get_mut(&mut self) -> ElemBorrowMut {
         return self.x.borrow_mut();
     }
 }
 
 
-pub struct ElementReq {
-    pub x_req: LReq,
-    pub y_req: LReq,
-}
-
-impl ElementReq {
-    pub fn new() -> ElementReq {
-        return ElementReq{x_req: LReq::new_empty(), y_req: LReq::new_empty()};
-    }
-
-    pub fn new_from_reqs(x_req: LReq, y_req: LReq) -> ElementReq {
-        return ElementReq{x_req: x_req, y_req: y_req};
-    }
-}
-
-
-pub struct ElementAlloc {
-    pub x_alloc: LAlloc,
-    pub y_alloc: LAlloc,
-    pub layout_flags: u8,
-}
-
-impl ElementAlloc {
-    pub fn new() -> ElementAlloc {
-        return ElementAlloc{x_alloc: LAlloc::new_empty(), y_alloc: LAlloc::new_empty(),
-                            layout_flags: LAYOUT_FLAGS_ALL_CLEAN};
-    }
-}
-
-
-pub trait TElementLayout {
+pub trait TElement {
+    /// Acquire reference to the element layout requisition
     fn element_req(&self) -> &ElementReq;
+    /// Acquire reference to the element layout allocation
     fn element_alloc(&self) -> &ElementAlloc;
+    /// Acquire reference to element layout requisition and mutable allocation
     fn element_req_and_mut_alloc(&mut self) -> (&ElementReq, &mut ElementAlloc);
 
+    /// Acquire reference to element layout X requisition
     fn x_req(&self) -> &LReq {
         return &self.element_req().x_req;
     }
 
+    /// Acquire reference to element layout X allocation
     fn x_alloc(&self) -> &LAlloc {
         return &self.element_alloc().x_alloc;
     }
 
+    /// Acquire reference to element layout X requisition and mutable X allocation
     fn x_req_and_mut_alloc(&mut self) -> (&LReq, &mut LAlloc) {
         let ra = self.element_req_and_mut_alloc();
         return (&ra.0.x_req, &mut ra.1.x_alloc);
     }
 
+    /// Acquire reference to element layout Y requisition
     fn y_req(&self) -> &LReq {
         return &self.element_req().y_req;
     }
 
+    /// Acquire reference to element layout Y allocation
     fn y_alloc(&self) -> &LAlloc {
         return &self.element_alloc().y_alloc;
     }
 
+    /// Acquire reference to element layout Y requisition and mutable Y allocation
     fn y_req_and_mut_alloc(&mut self) -> (&LReq, &mut LAlloc) {
         let ra = self.element_req_and_mut_alloc();
         return (&ra.0.y_req, &mut ra.1.y_alloc);
     }
-}
 
 
-pub trait TElement : TElementLayout {
+    /// Paint content contributed by `self`
```

Comment:
Are these methods with empty bodies supposed to say "the default is do nothing" or "you must implement this yourself"? If the latter, should they just be `fn draw_self(...);` without a body?